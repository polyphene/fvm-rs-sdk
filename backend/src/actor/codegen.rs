//! Codegen has the logic of code generation for our actor through the `#[fvm_state]` macro.

use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use std::convert::TryInto;

use crate::ast;
use crate::ast::Mutability;

impl ToTokens for ast::ActorImplementation {
    fn to_tokens(&self, into: &mut TokenStream) {
        let impl_member = self.rust_name.clone();
        let mut entry_points: Vec<TokenStream> = vec![];

        for entry_point in self.entry_points.iter() {
            let entry_point_value: u64 = entry_point.binding.clone().try_into().unwrap();
            let method_name = entry_point.rust_name.clone();

            let mut method_call = TokenStream::new();

            // If method not pure load state
            if !entry_point.mutability.is_pure() {
                // let keyword
                quote!(let).to_tokens(&mut method_call);
                if entry_point.mutability.is_write() {
                    // Add mut keyword if write mutability
                    quote!(mut).to_tokens(&mut method_call);
                }
                // Finalize load() call
                quote!(
                    state = #impl_member::load();
                )
                .to_tokens(&mut method_call);
            }

            // If method returns then store method result in variable
            if entry_point.returns {
                quote!(
                    let method_return =
                )
                .to_tokens(&mut method_call);
            }

            // Handle method calling based on mutability
            match entry_point.mutability {
                Mutability::Pure => quote!(
                    #impl_member::#method_name();
                )
                .to_tokens(&mut method_call),
                Mutability::View => quote!(
                    state.#method_name();
                )
                .to_tokens(&mut method_call),
                Mutability::Write => quote!(
                    state.#method_name();
                )
                .to_tokens(&mut method_call),
            };

            // If method returns then convert result to bytes
            if entry_point.returns {
                quote!(
                    ret = match(fvm_rs_sdk::encoding::to_vec(&method_return)) {
                        Ok(ret) => Some(fvm_rs_sdk::encoding::RawBytes::new(ret)),
                        Err(err) => {
                            fvm_rs_sdk::syscall::vm::abort(
                                fvm_rs_sdk::shared::error::ExitCode::USR_ILLEGAL_STATE.value(),
                                Some(format!("failed to serialize return value: {:?}", err).as_str()),
                            );
                        }
                    };
                ).to_tokens(&mut method_call)
            }

            // If mutability is write then save state
            if entry_point.mutability.is_write() {
                quote!(
                    state.save();
                )
                .to_tokens(&mut method_call);
            }

            entry_points.push(quote!(
                #entry_point_value => {
                    let mut ret = None;
                    #method_call
                    ret
                }
            ));
        }

        quote!(
            #[no_mangle]
            pub fn invoke(params_pointer: u32) -> u32 {
                // Conduct method dispatch. Handle input parameters and return data.
                let ret: Option<fvm_rs_sdk::encoding::RawBytes> =
                    match fvm_rs_sdk::syscall::message::method_number() {
                        #(#entry_points),*
                        _ => fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_UNHANDLED_MESSAGE.value(),
                            Some("unrecognized method"),
                        ),
                    };

                match ret {
                    None => fvm_rs_sdk::syscall::NO_DATA_BLOCK_ID,
                    Some(v) => match fvm_rs_sdk::syscall::ipld::put_block(
                        fvm_rs_sdk::encoding::DAG_CBOR,
                        v.bytes(),
                    ) {
                        Ok(id) => id,
                        Err(err) => fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_SERIALIZATION.value(),
                            Some(format!("failed to store return value: {}", err).as_str()),
                        ),
                    },
                }
            }
        )
        .to_tokens(into)
    }
}

#[cfg(test)]
mod tests {}
