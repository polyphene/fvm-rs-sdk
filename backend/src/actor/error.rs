#[derive(thiserror::Error, Debug)]
/// Errors related to an actor's interface.
pub enum Error {
    /// This error is thrown when the specified attribute is not handled
    #[error("unknown attribute '{0}'")]
    UnknownAttribute(String),
    /// This error is thrown when the specified dispatch method is not handled
    #[error("unknown dispatch method '{0}'")]
    UnkownDispatchMethod(String),
    /// This error is thrown when the dispatch method is not a literal string
    #[error("invalid codec format, {0}")]
    InvalidDispatchMethodFormat(String),
}
