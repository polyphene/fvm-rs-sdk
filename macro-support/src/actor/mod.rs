pub(crate) mod attrs;
mod error;
pub(crate) mod parser;
