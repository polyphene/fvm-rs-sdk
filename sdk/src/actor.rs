pub use fvm_rs_sdk_macro::fvm_actor;
