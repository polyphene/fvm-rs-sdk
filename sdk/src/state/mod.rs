//! State contains necessary code to handle a state object in an actor
pub mod cbor;
mod error;

use cid::Cid;
pub use fvm_ipld_blockstore::{Block, Blockstore};
pub use fvm_rs_sdk_macro::fvm_state;

/// StateObject is a trait to read and write an actor's state on the Filecoin Virtual Machine
pub trait StateObject {
    // Load state object from the FVM state
    fn load() -> Self;
    // Save object as an actor's state
    fn save(&self) -> Cid;
}
