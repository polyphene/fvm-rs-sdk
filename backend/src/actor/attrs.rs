//! Contains attributes available for the `#[fvm_actor]` procedural macro.

use std::convert::TryFrom;

use crate::utils::AnyIdent;
use anyhow::Result;
use syn::parse::{Parse, ParseStream, Result as SynResult};

use crate::actor::error::Error::{
    InvalidDispatchMethodFormat, UnknownAttribute, UnkownDispatchMethod,
};

#[derive(Clone, Debug)]
pub enum ActorAttr {
    Dispatch(Dispatch),
}

impl TryFrom<String> for ActorAttr {
    type Error = crate::actor::error::Error;

    fn try_from(attr: String) -> Result<Self, Self::Error> {
        match attr.as_str() {
            "dispatch" => Ok(ActorAttr::Dispatch(Dispatch::default())),
            _ => Err(UnknownAttribute(attr)),
        }
    }
}

impl Parse for ActorAttr {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let original = input.fork();
        let attr: AnyIdent = input.parse()?;
        let attr = attr.0;

        match ActorAttr::try_from(attr.to_string()) {
            Ok(ActorAttr::Dispatch(_)) => {
                input.parse::<syn::token::Eq>()?;
                let val = match input.parse::<syn::LitStr>() {
                    Ok(str) => match Dispatch::try_from(str.value()) {
                        Ok(state_attr) => state_attr,
                        Err(err) => return Err(original.error(format!("{}", err))),
                    },
                    Err(err) => {
                        return Err(original
                            .error(format!("{}", InvalidDispatchMethodFormat(err.to_string()))))
                    }
                };
                Ok(ActorAttr::Dispatch(val))
            }
            Err(err) => Err(original.error(format!("{}", err))),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Dispatch {
    Numeric,
}

impl Default for Dispatch {
    fn default() -> Self {
        Dispatch::Numeric
    }
}

impl TryFrom<String> for Dispatch {
    type Error = crate::actor::error::Error;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        match value.as_str() {
            "method-num" => Ok(Dispatch::Numeric),
            _ => Err(UnkownDispatchMethod(value)),
        }
    }
}
