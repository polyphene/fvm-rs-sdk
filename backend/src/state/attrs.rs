//! Contains attributes available for the `#[fvm_state]` procedural macro.

use std::convert::TryFrom;

use crate::utils::AnyIdent;
use anyhow::Result;
use syn::parse::{Parse, ParseStream, Result as SynResult};

use crate::state::error::Error::{InvalidCodecFormat, UnknownAttribute, UnknownCodec};

#[derive(Clone, Debug)]
pub enum StateAttr {
    Codec(Codec),
}

impl TryFrom<String> for StateAttr {
    type Error = crate::state::error::Error;

    fn try_from(attr: String) -> Result<Self, Self::Error> {
        match attr.as_str() {
            "codec" => Ok(StateAttr::Codec(Codec::default())),
            _ => Err(UnknownAttribute(attr)),
        }
    }
}

impl Parse for StateAttr {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let original = input.fork();
        let attr: AnyIdent = input.parse()?;
        let attr = attr.0;

        return match StateAttr::try_from(attr.to_string()) {
            Ok(StateAttr::Codec(_)) => {
                input.parse::<syn::token::Eq>()?;
                let val = match input.parse::<syn::LitStr>() {
                    Ok(str) => match Codec::try_from(str.value()) {
                        Ok(state_attr) => state_attr,
                        Err(err) => return Err(original.error(format!("{}", err))),
                    },
                    Err(err) => {
                        return Err(
                            original.error(format!("{}", InvalidCodecFormat(err.to_string())))
                        )
                    }
                };
                Ok(StateAttr::Codec(val))
            }
            Err(err) => Err(original.error(format!("{}", err))),
        };
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Codec {
    DagCbor,
}

impl Default for Codec {
    fn default() -> Self {
        Codec::DagCbor
    }
}

impl TryFrom<String> for Codec {
    type Error = crate::state::error::Error;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        match value.as_str() {
            "dag-cbor" => Ok(Codec::DagCbor),
            _ => Err(UnknownCodec(value)),
        }
    }
}
