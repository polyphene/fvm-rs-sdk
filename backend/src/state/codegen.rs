//! Codegen has the logic of code generation for our actor through the `#[fvm_state]` macro.

use proc_macro2::TokenStream;
use quote::{quote, ToTokens};

use crate::{ast, Diagnostic};

/// A trait for converting AST structs into Tokens and adding them to a TokenStream,
/// or providing a diagnostic if conversion fails.
pub trait TryToTokens {
    /// Attempt to convert a `Self` into tokens and add it to the `TokenStream`
    fn try_to_tokens(&self, into: &mut TokenStream) -> Result<(), Diagnostic>;

    /// Attempt to convert a `Self` into a new `TokenStream`
    fn try_to_token_stream(&self) -> Result<TokenStream, Diagnostic> {
        let mut tokens = TokenStream::new();
        self.try_to_tokens(&mut tokens)?;
        Ok(tokens)
    }
}

impl TryToTokens for ast::Program {
    // Generate wrappers for all the items that we've found
    fn try_to_tokens(&self, into: &mut TokenStream) -> Result<(), Diagnostic> {
        // Handling tagged structures
        for s in self.state_structs.iter() {
            s.to_tokens(into);
        }

        Ok(())
    }
}

impl ToTokens for ast::StateStruct {
    fn to_tokens(&self, into: &mut TokenStream) {
        // Add derive for serialize & deserialize
        *into = (quote! {
            #[derive(fvm_rs_sdk::encoding::tuple::Serialize_tuple, fvm_rs_sdk::encoding::tuple::Deserialize_tuple)]
            #[serde( crate = "fvm_rs_sdk::encoding::serde")]
            #into
        })
            .to_token_stream();

        self.generate_state_interface().to_tokens(into)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::attrs::Codec::DagCbor;

    #[test]
    fn basic_struct() {
        // Instantiate expected result
        let mut expected_final_stream = TokenStream::new();

        (quote! {
            #[derive(fvm_rs_sdk::encoding::tuple::Serialize_tuple, fvm_rs_sdk::encoding::tuple::Deserialize_tuple)]
            #[serde( crate = "fvm_rs_sdk::serde")]
            pub struct MockStruct {
                pub count: u64
            }

            impl fvm_rs_sdk::state::StateObject for MockStruct {
                fn load() -> Self {
                    use fvm_rs_sdk::state::Blockstore;

                    // First, load the current state root.
                    let root = match fvm_rs_sdk::syscall::sself::root() {
                        Ok(root) => root,
                        Err(err) => fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_ILLEGAL_STATE.value(),
                            Some(format!("failed to get root: {:?}", err).as_str()),
                        ),
                    };

                    // Get state's bytes
                    match fvm_rs_sdk::state::cbor::CborBlockstore.get(&root) {
                        // State fetched, convert byte to actual struct
                        Ok(Some(state_bytes)) => match fvm_rs_sdk::encoding::from_slice(&state_bytes) {
                            Ok(state) => state,
                            Err(err) => fvm_rs_sdk::syscall::vm::abort(
                                fvm_rs_sdk::shared::error::ExitCode::USR_SERIALIZATION.value(),
                                Some(format!("failed to deserialize state: {}", err).as_str()),
                            )
                        },
                        // No state
                        Ok(None) => fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_ILLEGAL_STATE.value(),
                            Some("state does not exist"),
                        ),
                        // Fetching failed
                        Err(err) => fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_ILLEGAL_STATE.value(),
                            Some(format!("failed to get state: {}", err).as_str()),
                        ),
                    }
                }

                fn save(&self) -> fvm_rs_sdk::cid::Cid {
                    use fvm_rs_sdk::state::Blockstore;

                    // Serialize state
                    let serialized = match fvm_rs_sdk::encoding::to_vec(self) {
                        Ok(s) => s,
                        Err(err) => fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_SERIALIZATION.value(),
                            Some(format!("failed to serialize state: {:?}", err).as_str()),
                        ),
                    };

                    // Put state
                    let cid = match fvm_rs_sdk::state::cbor::CborBlockstore.put(
                        fvm_rs_sdk::cid::Code::Blake2b256.into(),
                        &fvm_rs_sdk::state::Block {
                            codec: fvm_rs_sdk::encoding::DAG_CBOR,
                            data: serialized.as_slice()
                        }
                    ) {
                        Ok(cid) => cid,
                        Err(err) => fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_SERIALIZATION.value(),
                            Some(format!("failed to store initial state: {:}", err).as_str()),
                        ),
                    };
                    if let Err(err) = fvm_rs_sdk::syscall::sself::set_root(&cid) {
                        fvm_rs_sdk::syscall::vm::abort(
                            fvm_rs_sdk::shared::error::ExitCode::USR_ILLEGAL_STATE.value(),
                            Some(format!("failed to set root cid: {:}", err).as_str()),
                        );
                    }
                    cid
                }
            }
        })
            .to_tokens(&mut expected_final_stream);

        // Create new token stream
        let mut token_stream = TokenStream::new();

        // Add a structure to our stream
        (quote! {
            pub struct MockStruct {
                pub count: u64
            }
        })
        .to_tokens(&mut token_stream);

        // Convert token stream to syn structure
        let item = syn::parse2::<syn::Item>(token_stream.clone()).unwrap();

        match item {
            syn::Item::Struct(mut s) => {
                // Generate ast::Struct
                let mut fields = Vec::new();
                for field in s.fields.iter_mut() {
                    // Derive field name from ident
                    let (name, member) = match &field.ident {
                        Some(ident) => (ident.to_string(), syn::Member::Named(ident.clone())),
                        _ => unreachable!(),
                    };

                    fields.push(ast::StateStructField {
                        rust_name: member,
                        name,
                        struct_name: s.ident.clone(),
                        ty: field.ty.clone(),
                    });
                }

                let ast_struct = ast::StateStruct {
                    rust_name: s.ident.clone(),
                    name: s.ident.to_string(),
                    fields,
                    codec: DagCbor,
                };

                // Create ast::Program
                let program = ast::Program {
                    state_structs: vec![ast_struct],
                };

                program.try_to_tokens(&mut token_stream).unwrap();

                assert_eq!(token_stream.to_string(), expected_final_stream.to_string());
            }
            _ => unreachable!(),
        }
    }
}
