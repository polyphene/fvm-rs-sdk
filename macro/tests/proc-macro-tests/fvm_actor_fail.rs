#![allow(unreachable_code)]
use fvm_rs_sdk::actor::{fvm_actor, fvm_export};
use fvm_rs_sdk::state::*;

#[fvm_state]
pub struct MockStruct1 {
    pub count: u64,
}

// Fail because of generic
#[fvm_actor]
impl<T> MockStruct1 {
    #[fvm_export(binding = 1)]
    fn add(&mut self, a: T) {
        self.count += T.count
    }
}

// Fail because of ref
#[fvm_actor]
impl MockStruct1 {
    #[fvm_export(binding = 1)]
    pub fn add(&mut self, a: &u64) {
        self.count += a;
    }
}

// Fail because of bare function type
#[fvm_actor]
impl MockStruct1 {
    #[fvm_export(binding = 1)]
    pub fn add(&mut self, a: fn(u64) -> u64) {
        self.count += a
    }
}

// Fail because no binding
#[fvm_actor]
impl MockStruct1 {
    #[fvm_export]
    pub fn add(&mut self, a: u64) {
        self.count += a;
    }
}

// Fail because pointer type
#[fvm_actor]
impl MockStruct1 {
    #[fvm_export(binding = 1)]
    pub fn add(&mut self, a: *mut u64) {
        todo!()
    }
}

// Fail because private method
#[fvm_actor]
impl MockStruct1 {
    #[fvm_export(binding = 1)]
    fn add(&mut self, a: u64) {
        self.count += a;
    }
}

// Fail because never argument type
#[fvm_actor]
impl MockStruct1 {
    #[fvm_export(binding = 1)]
    pub fn call_never(&mut self, a: !) {
        todo!()
    }
}

// Fail because slice argument type
#[fvm_actor]
impl MockStruct1 {
    #[fvm_export(binding = 1)]
    pub fn call_never(&mut self, a: Box<[u64]>) {
        self.count += a[0]
    }
}

fn main() {}
