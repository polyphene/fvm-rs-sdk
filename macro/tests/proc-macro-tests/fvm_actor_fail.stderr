error: implementation with #[fvm_actor] cannot have lifetime or type parameters.
  --> tests/proc-macro-tests/fvm_actor_fail.rs:12:1
   |
12 | #[fvm_actor]
   | ^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `fvm_actor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: a referenced type, '& u64', can not be used as a type for an entry point argument.
  --> tests/proc-macro-tests/fvm_actor_fail.rs:21:1
   |
21 | #[fvm_actor]
   | ^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `fvm_actor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: a bare function type, 'fn(u64) -> u64', can not be used as a type for an entry point argument.
  --> tests/proc-macro-tests/fvm_actor_fail.rs:30:1
   |
30 | #[fvm_actor]
   | ^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `fvm_actor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected attribute arguments in parentheses: #[fvm_export(...)]
  --> tests/proc-macro-tests/fvm_actor_fail.rs:41:5
   |
41 |     #[fvm_export]
   |     ^^^^^^^^^^^^^

error: a pointer type, '* mut u64', can not be used as a type for an entry point argument.
  --> tests/proc-macro-tests/fvm_actor_fail.rs:48:1
   |
48 | #[fvm_actor]
   | ^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `fvm_actor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: the never type, '!', can not be used as a type for an entry point argument.
  --> tests/proc-macro-tests/fvm_actor_fail.rs:57:1
   |
57 | #[fvm_actor]
   | ^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `fvm_actor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: a slice type, '[u64]', can not be used as a type for an entry point argument.
  --> tests/proc-macro-tests/fvm_actor_fail.rs:66:1
   |
66 | #[fvm_actor]
   | ^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `fvm_actor` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unknown attribute 'mock'
  --> tests/proc-macro-tests/fvm_actor_fail.rs:77:18
   |
77 |     #[fvm_export(mock = 1)]
   |                  ^^^^

error: invalid 'method_num' value
  --> tests/proc-macro-tests/fvm_actor_fail.rs:86:18
   |
86 |     #[fvm_export(method_num = "hello")]
   |                  ^^^^^^^^^^
